ember.md

```diff
@@ -0,0 +1,461 @@
+
+1. __Plain properties__
+
+   Start with properties that configure the module's behavior. Examples are
+   `tagName` and `classNames` on components and `queryParams` on controllers and
+   routes. Followed by any other simple properties, like default values for properties.
+
+2. __Single line computed property macros__
+
+   E.g. `alias`, `sort` and other macros. Start with service injections. If the
+   module is a model, then `attr` properties should be first, followed by
+   `belongsTo` and `hasMany`.
+
+3. __Multi line computed property functions__
+
+4. __Lifecycle hooks__
+
+   The hooks should be chronologically ordered by the order they are invoked in.
+
+5. __Functions__
+
+   Public functions first, internal functions after.
+
+6. __Actions__
+
+```js
+export default Component.extend({
+  // Plain properties
+  tagName: 'span',
+
+  // Single line CP
+  post: alias('myPost'),
+
+  // Multiline CP
+  authorName: computed('author.{firstName,lastName}', function _authorName() {
+    // code
+  }),
+
+  // Lifecycle hooks
+  didReceiveAttrs() {
+    this._super(...arguments);
+    // code
+  },
+
+  // Functions
+  someFunction() {
+    // code
+  },
+
+  actions: {
+    someAction() {
+      // Code
+    },
+  },
+});
+```
+
+### Override init
+
+Rather than using the object's `init` hook via `on`, override init and
+call `_super` with `...arguments`. This allows you to control execution
+order. [Don't Don't Override Init](https://dockyard.com/blog/2015/10/19/2015-dont-dont-override-init)
+
+## Models
+
+### Organization
+
+Models should be grouped as follows:
+
+* Attributes
+* Associations
+* Computed Properties
+
+Within each section, the attributes should be ordered alphabetically.
+
+```js
+// Good
+
+import Ember from 'ember';
+import DS from 'ember-data';
+
+const { computed } = Ember;
+
+const {
+  attr,
+  hasMany,
+} = DS;
+
+export default DS.Model.extend({
+  // Attributes
+  firstName: attr('string'),
+  lastName: attr('string'),
+
+  // Associations
+  children: hasMany('child'),
+
+  // Computed Properties
+  fullName: computed('firstName', 'lastName', function _fullName() {
+    // Code
+  }),
+});
+
+// Bad
+
+import Ember from 'ember';
+import DS from 'ember-data';
+
+const { computed } = Ember;
+
+const {
+  attr,
+  hasMany,
+} = DS;
+
+export default DS.Model.extend({
+  children: hasMany('child'),
+  firstName: attr('string'),
+  lastName: attr('string'),
+
+  fullName: computed('firstName', 'lastName', function _fullName() {
+    // Code
+  }),
+});
+
+```
+
+## Controllers
+
+### Define query params first
+
+For consistency and ease of discover, list your query params first in
+your controller. These should be listed above default values.
+
+### Alias your model
+
+It provides a cleaner code to name your model `user` if it is a user. It
+is more maintainable, and will fall in line with future routable
+components
+
+```javascript
+export default Ember.Controller.extend({
+  user: computed.readOnly('model'),
```

Comment:
I was wondering about this. I'm not sure what the best practice should be. I'm tempted to say that everything should actually be `readOnly` unless it needs to be injected.

Here's a little bit of background.

I started using `alias` for everything, as it seemed like the obvious thing that I was doing. However, since alias passes through setting values as well as getting, it allows changes to propagate which is something the Ember community learned early on can cause bugs that are hard to troubleshot. I won't go in to too much detail here, but that's why the switch to data-down-actions-up over two-way bindings.

For more on those concepts, check out https://balinterdi.com/blog/the-tale-of-two-bindings/ and http://www.samselikoff.com/blog/data-down-actions-up/.

So, I switched to using `readOnly` as a default for these basic aliases. `readOnly` is great because it will cause an exception when the property is set, which is great if you are not expecting the property to be set.

However, sometimes you want to inject values for these `readOnly` properties, which `oneWay` allows you to do. So, I switched to using `oneWay` as a default. However, I'm wondering if we should all use `readOnly` as a default until we need it to be more permissive. By using the strictest as the default, we prevent issues where a property is set that was not expected to be set. I actually ran in to an issue like this in topic-reports. The downside of this is that people would need to change values to `oneWay` from `readOnly` for test injection (as you mentioned). So, `oneWay` has been the default, but I'm actually a little concerned with using it as a default now.

I want to point out here that I'm only talking about a default. `readOnly`, `oneWay`, and `alias` can all be used if necessary. For example, maybe you do need `alias` because having the setter propagate is important. However, using an action should probably be preferred.

Anyway, that's a bit of rambling. I'm not sure we know the full implications of picking one of these right now.

Does anyone else have thoughts on this?